prow/git/v2/interactor.go

```diff
@@ -52,3 +62,262 @@ type Interactor interface {
+}
+
+// Only the `merge` and `squash` strategies are supported.
+func (i *interactor) MergeAndCheckout(baseSHA string, mergeStrategy string, headSHAs ...string) error {
+	if baseSHA == "" {
+		return errors.New("baseSHA must be set")
+	}
+	if err := i.Checkout(baseSHA); err != nil {
+		return err
+	}
+	for _, headSHA := range headSHAs {
+		ok, err := i.MergeWithStrategy(headSHA, mergeStrategy)
+		if err != nil {
+			return err
+		} else if !ok {
+			return fmt.Errorf("failed to merge %q", headSHA)
+		}
+	}
+	return nil
+}
+
+// Am tries to apply the patch in the given path into the current branch
+// by performing a three-way merge (similar to git cherry-pick). It returns
+// an error if the patch cannot be applied.
+func (i *interactor) Am(path string) error {
+	i.logger.Infof("Applying patch at %s", path)
+	out, err := i.executor.Run("am", "--3way", path)
+	if err == nil {
+		return nil
+	}
+	i.logger.WithError(err).Infof("Patch apply failed with output: %s", string(out))
+	if abortOut, abortErr := i.executor.Run("am", "--abort"); err != nil {
+		i.logger.WithError(abortErr).Warningf("Aborting patch apply failed with output: %s", string(abortOut))
+	}
+	return errors.New(string(bytes.TrimPrefix(out, []byte("The copy of the patch that failed is found in: .git/rebase-apply/patch"))))
+}
+
+// RemoteUpdate fetches all updates from the remote.
+func (i *interactor) RemoteUpdate() error {
+	i.logger.Info("Updating from remote")
+	if out, err := i.executor.Run("remote", "update"); err != nil {
+		return fmt.Errorf("error updating: %v %v", err, string(out))
+	}
+	return nil
+}
+
+// Fetch fetches all updates from the remote.
+func (i *interactor) Fetch() error {
+	remote, err := i.remote()
+	if err != nil {
+		return fmt.Errorf("could not resolve remote for fetching: %v", err)
+	}
+	i.logger.Infof("Fetching from %s", remote)
+	if out, err := i.executor.Run("fetch", remote); err != nil {
+		return fmt.Errorf("error fetching: %v %v", err, string(out))
+	}
+	return nil
+}
+
+// FetchRef fetches a refspec from the remote and leaves it as FETCH_HEAD.
+func (i *interactor) FetchRef(refspec string) error {
+	remote, err := i.remote()
+	if err != nil {
+		return fmt.Errorf("could not resolve remote for fetching: %v", err)
+	}
+	i.logger.Infof("Fetching %q from %s", refspec, remote)
+	if out, err := i.executor.Run("fetch", remote, refspec); err != nil {
+		return fmt.Errorf("error fetching %q: %v %v", refspec, err, string(out))
+	}
+	return nil
+}
+
+// CheckoutPullRequest fetches the HEAD of a pull request using a synthetic refspec
+// available on GitHub remotes and creates a branch at that commit.
+func (i *interactor) CheckoutPullRequest(number int) error {
+	i.logger.Infof("Checking out pull request %d", number)
+	if err := i.FetchRef(fmt.Sprintf("pull/%d/head", number)); err != nil {
+		return err
+	}
+	if err := i.Checkout("FETCH_HEAD"); err != nil {
+		return err
+	}
+	if err := i.CheckoutNewBranch(fmt.Sprintf("pull%d", number)); err != nil {
```

Comment:
This will make `CheckoutPullRequest` fail if called multiple times - so it should be documented, or `CheckoutNewBranch` should be made to allow multiple calls for the same branch name.

Reply:
This simply achieves parity with the old client. I don't think we should change the way this works in this PR. Follow-ups after the `v1` client is removed could change this.